pkg/agent/informant.go

```diff
@@ -254,6 +255,14 @@ func NewInformantServer(
 		}
 	})
 
+	// Thread waiting for the context to be canceled so we can use it to shut down the server
+	shutdownWaiterName := fmt.Sprintf("InformantServer shutdown waiter (%s)", server.desc.AgentID)
+	runner.spawnBackgroundWorker(ctx, shutdownWaiterName, func(c context.Context) {
+		// Wait until parent context OR server's context is done.
+		<-backgroundCtx.Done()
```

Comment:
it looks like the way this gets canceled is if `server.exit` runs, which this function then calls in the next line (am I reading this correctly?) I suppose other things call exit too.) 

If the parent context gets canceled (e.g. the one passed to spawnBackgroundWorker) will this run? I suppose it will because nothing is waiting on `c` (I might just delete that variable in the signature to make that more clear). 

it seems plausible that this could help, but it's not totally clear to me that this is *definitely* the right solution: moving L204-213 out of the exit function blocking on background/the input context being canceled seems _roughly_ equally functional. Thought I don't have a perfect mental model for all of the threads and where the informant monitoring parts in the agent can exit, so if this is the best way, then let's do it?

Reply:
> it looks like the way this gets canceled is if server.exit runs, which this function then calls in the next line (am I reading this correctly?) I suppose other things call exit too.)
>
> If the parent context gets canceled (e.g. the one passed to spawnBackgroundWorker) will this run?

Yeah so `server.exit` is called from a handful of places, _usually_ due to some error. (before this PR, _only_ due to an error) `backgroundCtx` is derived from the parent context, so all should be good on that front.

> (I might just delete that variable in the signature to make that more clear)

done

> moving L204-213 out of the exit function blocking on background/the input context being canceled seems roughly equally functional

At first I was about to agree & implement this. However: other parts of the code rely on `server.exitStatus != nil` being equivalent to "the server is no longer running", and `server.exit` sets `exitStatus`, so avoid it would create weird behavior when the context is canceled.

That said, maybe the semantics there are wrong.